### PR TITLE
kubeadm: add support for skipping phases in v1beta3

### LINKED ISSUE #2261: support skipping phases via the kubeadm API
(i though we had a ticket about this already, but i cannot find it?)

there have been a few requests about support for skipping phases via the kubeadm config.
one was from stealthybox related to addon support and the other one i saw was here from CAPI:
https://github.com/kubernetes-sigs/cluster-api/issues/2769#issuecomment-681100776

these can be string slices under JoinConfiguration and InitConfiguration.
for the future "upgrade apply" phases, we don't have UpgradeConfiguration...

### PR DESCRIPTION
**Pull request by neolit123**
#### What this PR does / why we need it:

- Add the field SkipPhases in the public v1beta3 as a []string (omitempty)
- Add the field in the internal type
- Run generators
- Adapt v1beta2 converter for JoinConfiguration

During Runner data initialization, if the value for the flag
"--skip-phases" was empty set the {init|join}Runner.Options.SkipPhases
to the {Init|Join}Configuration.SkipPhases value.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/2261
xref https://github.com/kubernetes/kubeadm/issues/1796

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: add a new field "skipPhases" to v1beta3 InitConfiguration and JoinConfiguration that can contain a list of phases to skip during "kubeadm init" and "kubeadm join". The flag "--skip-phases" takes precedence over this field.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/970-kubeadm-config
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -193,7 +193,15 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
...
 	initRunner.SetDataInitializer(func(cmd *cobra.Command, args []string) (workflow.RunData, error) {
-		return newInitData(cmd, args, initOptions, out)
+		data, err := newInitData(cmd, args, initOptions, out)
+		if err != nil {
+			return nil, err
+		}
+		// If the flag for skipping phases was empty, use the values from config
+		if len(initRunner.Options.SkipPhases) == 0 {
+			initRunner.Options.SkipPhases = data.cfg.SkipPhases
+		}
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/init.go:203]: This is one of the few cases in kubeadm where we are assuming that flags and config could be set at the same time for the same field and we are enforcing a precedence rule (flags over config).
This leads me to two considerations/questions:
- Should me make this behavior more clear by documenting it both in the API and on the flags?
- Are we sure this behaviour (flag over config) is consistent across all the flags that can co-exists with config?

**Review by fabriziopandini — COMMENTED** [review]
Overall changes looks good to me.
My only concern is about the transient nature of InitConfigurations and JoinConfiguration, which mirrors the imperative nature of kubeadm commands, but doesn't really helps in the management of the cluster lifecycle.

To make things a little bit more concrete, let's assume the user uses those new fields for skipping addons installation. Everything works, but there is no way to depict a cluster without addons from a cluster with addons unless you inspect the cluster (which is something kubeadm itself is already doing e.g. for making upgrades to work without addons installed).

Wondering if we can better address this...
@yastij @vincepri opinions?

**Comment by neolit123** [comment]
> To make things a little bit more concrete, let's assume the user uses those new fields for skipping addons installation. Everything works, but there is no way to depict a cluster without addons from a cluster with addons unless you inspect the cluster (which is something kubeadm itself is already doing e.g. for making upgrades to work without addons installed).

this is also a problem where a joining node, has no knowledge about the existing nodes or cluster state.
the higher level user / tool has control over the initial cluster state and what will land on that cluster.
i.e. they know the contents of the InitConfiguration and different JoinConfiguration.
they can also decide to store these in config maps.

the missing peace for kubeadm mutable upgrades and kubeadm upgrade apply are:
- phases
- UpgradeConfiguration of sorts, so that upgrade supports not only CLI flags.

today "kubeadm upgrade" is automatic in a limited way.

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/cmd/init.go:203]: i'm hoping that long term we can remove flags like --skip-phases. and allow only usage from config.

> Should me make this behavior more clear by documenting it both in the API and on the flags?

i can document it in both places.

> Are we sure this behaviour (flag over config) is consistent across all the flags that can co-exists with config?

yes, one exception is --cri-socket, which for 1.22 we are disallowing as mixed with --config.
another "long term" note here is that criSocket must be a field in KubeletConfiguration and kubeadm should not care about it. have no, crictl vs docker CLI, abstraction and so on.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/init.go:203]: added more comments.

**Comment by neolit123** [comment]
@fabriziopandini 
this is ready for another pass / LGTM.

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 2]
```diff
@@ -313,6 +318,11 @@ type JoinConfiguration struct {
 	// ControlPlane defines the additional control plane instance to be deployed on the joining node.
 	// If nil, no additional control plane instance will be deployed.
 	ControlPlane *JoinControlPlane
+
+	// SkipPhases is a list of phases to skip during command execution.
+	// The list of phases can be obtained with the "kubeadm init --help" command.
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/types.go:323]: Should this be 
```suggestion
	// The list of phases can be obtained with the "kubeadm join --help" command.
```

**Review by fabriziopandini — COMMENTED** [review]
lgtm pending the last nit in comments

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: updated.

**Pull request merged**